kvbc/src/categorization/immutable_kv_category.cpp

```diff
@@ -120,14 +131,100 @@ void ImmutableKeyValueCategory::deleteBlock(const ImmutableUpdatesInfo &updates_
   }
 }
 
-std::optional<ImmutableValue> ImmutableKeyValueCategory::get(const std::string &key) const {
-  const auto ser = db_->get(cf_, key);
+std::optional<Value> ImmutableKeyValueCategory::get(const std::string &key, BlockId block_id) const {
+  auto val = getLatest(key);
+  if (!val) {
+    return std::nullopt;
+  }
+  if (asImmutable(val).block_id != block_id) {
+    return std::nullopt;
+  }
+  return val;
+}
+
+std::optional<Value> ImmutableKeyValueCategory::getLatest(const std::string &key) const {
+  const auto ser = db_->getSlice(cf_, key);
   if (!ser) {
     return std::nullopt;
   }
-  auto value = ImmutableValue{};
-  deserialize(*ser, value);
-  return value;
+  return value(*ser);
+}
+
+void ImmutableKeyValueCategory::multiGet(const std::vector<std::string> &keys,
+                                         const std::vector<BlockId> &versions,
+                                         std::vector<std::optional<Value>> &values) const {
+  ConcordAssertEQ(keys.size(), versions.size());
+
+  auto slices = std::vector<::rocksdb::PinnableSlice>{};
+  auto statuses = std::vector<::rocksdb::Status>{};
+  db_->multiGet(cf_, keys, slices, statuses);
+
+  values.clear();
+  for (auto i = 0ull; i < slices.size(); ++i) {
+    const auto &status = statuses[i];
+    const auto &slice = slices[i];
+    const auto version = versions[i];
+    if (status.ok()) {
+      const auto v = value(slice);
+      if (v.block_id == version) {
+        values.push_back(v);
+      } else {
+        values.push_back(std::nullopt);
+      }
+    } else if (status.IsNotFound()) {
+      values.push_back(std::nullopt);
+    } else {
+      throw std::runtime_error{"ImmutableKeyValueCategory multiGet() failure: " + status.ToString()};
+    }
```

Comment:
nit : handling non-ok fist and reduce nesting seems clearer to me
```
if(status.IsNotFound()){
     values.push_back(std::nullopt);
     continue;
}

if(!status.ok()){
    throw std::runtime_error{"ImmutableKeyValueCategory multiGet() failure: " + status.ToString()};
}

const auto v = value(slice);
if (v.block_id != version) {
  values.push_back(std::nullopt);
  continue;
} 

values.push_back(v);
}

Reply:
Not sure if that matters for modern compilers and CPUs, but I think the `ok()` status is most common and we'd like to check for it first. I also wanted to avoid explicit `continue`. But I agree that nesting makes it a bit more complex. I will leave as is for now, if you are OK.

Reply:
I also dislike using continue, as it makes the control flow non-linear and harder to understand IMO.

Reply:
sure, just a nit offer